src/api_manager/context/request_context_test.cc

```diff
@@ -0,0 +1,393 @@
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+//
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+//#include "src/api_manager/mock_api_manager_environment.h"
+#include "src/api_manager/context/request_context.h"
+#include "src/api_manager/context/service_context.h"
+//#include "src/api_manager/mock_request.h"
+//#include "src/api_manager/api_manager_impl.h"
+#include "include/api_manager/request.h"
+
+using ::testing::_;
+using ::testing::Invoke;
+using ::testing::Mock;
+using ::testing::Return;
+
+using ::google::api_manager::utils::Status;
+
+namespace google {
+namespace api_manager {
+namespace context {
+
+namespace {
+
+const char kServerConfigWithoutClientIPExperiment[] =
+    R"(
+{
+  "experimental": {
+    "client_real_ip_header": "X-Forwarded-For",
+    "disable_log_status": false
+  }
+}
+)";
+
+const char kServerConfigWithClientIPExperimentOutOfIndex[] =
+    R"(
+{
+  "real_client_ip_from_header_config": {
+    "client_real_ip_header": "X-Forwarded-For",
+    "client_real_ip_position": -5
+  }
+}
+)";
+
+const char kServerConfigWithClientIPExperimentSecondFromLast[] =
+    R"(
+{
+  "real_client_ip_from_header_config": {
+    "client_real_ip_header": "X-Forwarded-For",
+    "client_real_ip_position": -2
+  }
+}
+)";
+
+const char kServerConfigWithClientIPExperimentLast[] =
+    R"(
+{
+  "real_client_ip_from_header_config": {
+    "client_real_ip_header": "X-Forwarded-For",
+    "client_real_ip_position": -1
+  }
+}
+)";
+
+const char kServerConfigWithClientIPExperimentFirst[] =
+    R"(
+{
+  "real_client_ip_from_header_config": {
+    "client_real_ip_header": "X-Forwarded-For",
+    "client_real_ip_position": 0
+  }
+}
+)";
+
+const char kServerConfigWithClientIPExperimentSecond[] =
+    R"(
+{
+  "real_client_ip_from_header_config": {
+    "client_real_ip_header": "X-Forwarded-For",
+    "client_real_ip_position": 1
+  }
+}
+)";
+
+const char kServiceConfig1[] =
+    R"(
+{
+  "name": "bookstore.test.appspot.com",
+  "title": "Bookstore",
+  "http": {
+    "rules": [
+      {
+        "selector": "EchoGetMessage",
+        "get": "/echo"
+      }
+    ]
+  },
+  "usage": {
+    "rules": [
+      {
+        "selector": "EchoGetMessage",
+        "allowUnregisteredCalls": true
+      }
+    ]
+  },
+  "control": {
+    "environment": "servicecontrol.googleapis.com"
+  },
+  "id": "2017-05-01r0"
+}
+)";
+
+// Simulate periodic timer event on creation
+class MockPeriodicTimer : public PeriodicTimer {
+ public:
+  MockPeriodicTimer() {}
+  MockPeriodicTimer(std::function<void()> continuation)
+      : continuation_(continuation) {
+    continuation_();
+  }
+
+  virtual ~MockPeriodicTimer() {}
+  void Stop() {}
+
+ private:
+  std::function<void()> continuation_;
+};
+
+class MockApiManagerEnvironment : public ApiManagerEnvInterface {
```

Comment:
the whole test should be called client_ip_extraction_test

Reply:
Renamed to client_ip_extraction_test